File: packages/typescript-sdk/src/identity.ts
import { z } from "zod";

const AGENT_PREFIX = "agent:";
const USER_PREFIX = "user:";

export const PrincipalTypeSchema = z.enum(["user", "agent"]);
export type PrincipalType = z.infer<typeof PrincipalTypeSchema>;

export interface SessionCreatorPrincipal {
  tenant_id: string;
  id: string;
  type: PrincipalType;
}

export interface SessionTokenPrincipal {
  type: PrincipalType;
  id: string;
}

function hasPrincipalPrefix(id: string): boolean {
  return id.startsWith(AGENT_PREFIX) || id.startsWith(USER_PREFIX);
}

/**
 * Represents a resolved caller identity with tenant, principal id, and type.
 */
export class StarciteIdentity {
  readonly tenantId: string;
  readonly id: string;
  readonly type: PrincipalType;

  constructor(options: {
    tenantId: string;
    id: string;
    type: PrincipalType;
  }) {
    if (hasPrincipalPrefix(options.id)) {
      throw new Error(
        `StarciteIdentity id must not include a principal prefix; received '${options.id}'`
      );
    }

    this.tenantId = options.tenantId;
    this.id = options.id;
    this.type = options.type;
  }

  /**
   * Serializes to the `creator_principal` wire format expected by the API.
   */
  toCreatorPrincipal(): SessionCreatorPrincipal {
    return { tenant_id: this.tenantId, id: this.id, type: this.type };
  }

  /**
   * Serializes to the `principal` wire format used in session token requests.
   */
  toTokenPrincipal(): SessionTokenPrincipal {
    return { id: this.id, type: this.type };
  }

  /**
   * Returns the actor string derived from this identity (e.g. `agent:planner`, `user:alice`).
   */
  toActor(): string {
    return `${this.type}:${this.id}`;
  }
}

File: packages/python-sdk/src/starcite_sdk/identity.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

PrincipalType = Literal["user", "agent"]


@dataclass(frozen=True)
class StarciteIdentity:
    """Resolved caller identity scoped to one tenant."""

    tenant_id: str
    id: str
    type: PrincipalType

    def __post_init__(self) -> None:
        if self.id.startswith("agent:") or self.id.startswith("user:"):
            raise ValueError(
                f"StarciteIdentity id must not include a principal prefix; received '{self.id}'"
            )

    def to_creator_principal(self) -> dict[str, str]:
        return {"tenant_id": self.tenant_id, "id": self.id, "type": self.type}

    def to_token_principal(self) -> dict[str, str]:
        return {"id": self.id, "type": self.type}

    def to_actor(self) -> str:
        return f"{self.type}:{self.id}"
